### ISSUE TITLE
Assertion `shstrtabIndex < shdrs.size()' failed

### ISSUE THREAD

**Issue by pjones** [NixOS/patchelf#63]
I'm attempting to create a package for the [XCTU application](http://www.digi.com/products/xbee-rf-solutions/xctu-software/xctu) which is distributed only as a Linux binary.  Unfortunately, `patchelf` seems to barf on it:

```
patchelf: patchelf.cc:292: void ElfFile<Elf_Ehdr, Elf_Phdr, Elf_Shdr, Elf_Addr, Elf_Off, Elf_Dyn, Elf_Sym>::parse() [with Elf_Ehdr = Elf64_Ehdr; Elf_Phdr = Elf64_Phdr; Elf_Shdr = Elf64_Shdr; Elf_Addr = long unsigned int; Elf_Off = long unsigned int; Elf_Dyn = Elf64_Dyn; Elf_Sym = Elf64_Sym]: Assertion `shstrtabIndex < shdrs.size()' failed.
```

`file` says:

```
ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, stripped
```

`readelf -a` says:

```
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 03 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - GNU
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x46f9f0
  Start of program headers:          64 (bytes into file)
  Start of section headers:          0 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         3
  Size of section headers:           64 (bytes)
  Number of section headers:         0
  Section header string table index: 0

There are no sections in this file.

There are no sections to group in this file.

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000000 0x0000000000300000 0x0000000000300000
                 0x0000000000170b4e 0x0000000000170b4e  R E    100000
  LOAD           0x000000000014e508 0x000000000064e508 0x000000000064e508
                 0x0000000000000000 0x0000000000000000  RW     100000
  PAX_FLAGS      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000         4

There is no dynamic section in this file.

There are no relocations in this file.

The decoding of unwind sections for machine type Advanced Micro Devices X86-64 is not currently supported.

No version information found in this file.
```

**Comment by darealshinji** [comment]
I think using patchelf on static binaries doesn't make much sense. There's no dynamic section that could be modified ;-)

**Comment by pjones** [comment]
Ah, I forgot to mention that I'm primarily interested in using the `--set-interpreter` feature of `patchelf`.  That and `--print-interpreter` produce the same error message.  From `strace` I gathered that this static binary has the wrong interpreter:

~~~
open("/lib64/ld-linux-x86-64.so.2", O_RDONLY) = -1 ENOENT (No such file or directory)
~~~

I guess that means that the title of this issue might be incorrect.

**Comment by glines** [comment]
I'm having a similar problem with a completely different program: the [64 bit Linux installer for Teensyduino](http://www.pjrc.com/teensy/td_download.html).

The output from file and readelf is pretty much exactly the same as above. The strace is here:
```
28157 execve("./TeensyduinoInstall.linux64", ["./TeensyduinoInstall.linux64"], [/* 86 vars */]) = 0
28157 mmap(0xfba5000, 257558990, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, 0, 0) = 0xfba5000
28157 readlink("/proc/self/exe", "/tmp/TeensyduinoInstall.linux64", 4096) = 31
28157 mmap(0x400000, 259674112, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000
28157 mmap(0x400000, 763202, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000
28157 mprotect(0x400000, 763202, PROT_READ|PROT_EXEC) = 0
28157 mmap(0x6ba000, 256729620, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0xba000) = 0x6ba000
28157 mprotect(0x6ba000, 256729620, PROT_READ|PROT_WRITE) = 0
28157 mmap(0xfb91000, 79064, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0xfb91000
28157 open("/lib64/ld-linux-x86-64.so.2", O_RDONLY) = -1 ENOENT (No such file or directory)
28157 exit(127)                         = ?
28157 +++ exited with 127 +++
```
Clearly it is trying to open the dynamic linker, but neither patchelf nor anything else recognizes it as a dynamically linked executable. I also tried manually running the dynamic linker on the ./Teensyduino.linux64 file but it encountered a segfault.

I suspect that what it is doing is extracting itself and subsequently attempting to start the dynamic linker. It looks like XCTU uses InstallBuilder. I'm not sure what Teensyduino uses, but it's probably something similar.

I'm going to try to trick it into working by using FhsUserEnv to trick it into finding the dynamic linker. I'll report back whether that works or not.

**Comment by glines** [comment]
I got a tip from IRC that the Teensyduino installer happens to be a [UPX archive](https://upx.github.io/). I was able to extract the achive with `upx -d ./TeensyduinoInstall.linux64` and continue from there with the usual patchelf procedure.

Clearly the statically linked binary is extracting its contents and then calling the dynamic linker. I doubt patchelf could ever detect this situation, but it would be nice if it gave a better error message.
